docs/specification/draft/basic/primitive_authorization.mdx

```diff
@@ -0,0 +1,302 @@
+information disclosure about unauthorized capabilities.
+
+## Authorization Model
+
+### Primitive Classes
+
+MCP primitives can be classified into two authorization categories:
+
+| Primitive Class | Description                                                                                                             |
+| :-------------- | :---------------------------------------------------------------------------------------------------------------------- |
+| **Public**      | Anyone with access to the MCP server, whether authorized or not, has access to the primitive.                           |
+| **Restricted**  | To access the primitive, the MCP server must assess whether the requestor is authorized to access that specific entity. |
+
+### Authorization Metadata
+
+Each primitive definition **MAY** include an `authorization` component that specifies
+access control requirements. The presence or absence of this component determines
+the primitive's classification:
+
+- **Presence** of `authorization` component: The primitive requires specific authorization
+  evaluation and is not discoverable without satisfying the specified policy.
+- **Absence** of `authorization` component: The primitive may or may not require
+  global authorization as configured by the MCP server.
+
+<Warning>
+
+The `authorization` component is **exclusively for server-side policy evaluation**.
+Servers **MUST NOT** include the `authorization` field in any responses to clients.
+This metadata is used internally by the server to make access control decisions
+and must never be exposed to prevent information disclosure about authorization
+requirements.
+
+</Warning>
+
+## Primitive Authorization Structure
+
+### Example: Tool with Authorization
+
+```json
+{
+  "name": "create-file",
+  "description": "Creates a new file with the specified content at the given path",
+  "inputSchema": {
+    "type": "object",
+    "properties": {
+      "path": {
+        "type": "string",
+        "description": "The file path where the file should be created"
+      },
+      "content": {
+        "type": "string",
+        "description": "The content to write to the file"
+      }
+    },
+    "required": ["path", "content"]
+  },
+  "authorization": {
+    "allowed_roles": ["admin", "contributor", "manager"],
+    "allowed_scopes": ["files:write", "workspace:modify"],
+    "required_claims": {
+      "organization": "example-org"
+    }
+  },
+  "annotations": {
+    "title": "Create File",
+    "readOnlyHint": false,
+    "destructiveHint": true,
+    "idempotentHint": false,
+    "openWorldHint": true
+  }
+}
+```
+
+## Discovery and Access Control
+
+### Filtered Discovery
+
+When an MCP client invokes discovery operations (such as `tools/list`, `resources/list`,
+`prompts/list`, or `roots/list`), the server **MUST**:
+
+1. Extract authorization context from the request (e.g., `Authorization` header)
```

Comment:
This implies that the initial token that is presented for discovery needs to carry all the authorizations for the user/app interacting at the given time, breaking the rule of just-enough privileges. In an agentic setup, primitive access is dynamic based on the reasoning at the host. Better if the discovery is orthogonal to authorization evaluation.